File: scripts/generate-icon.py
#!/usr/bin/env python3
"""Draw the 1024 App Store icon using the live access: technology colours."""

from __future__ import annotations

import struct
import sys
from pathlib import Path

from PIL import Image, ImageDraw

CANVAS = 1024

# Colours from https://www.accesstechnology.co.uk
FIELD = (0x1A, 0x2C, 0x41)
SURFACE = (0xF5, 0xF7, 0xFA)
SHIFT = (0x24, 0x76, 0xED)
ACCENT = (0x33, 0xDE, 0xCF)

KEYBOARD = (168, 297, 688, 430)
KEY_SIZE = 48
GAP = 12
ROWS = (10, 9, 7)


def flip_rect(x: float, y: float, w: float, h: float) -> tuple[float, float, float, float]:
    return (x, CANVAS - y - h, w, h)


def rounded(draw: ImageDraw.ImageDraw, rect: tuple[float, float, float, float], radius: float, fill: tuple[int, int, int]) -> None:
    x, y, w, h = flip_rect(*rect)
    draw.rounded_rectangle((x, y, x + w, y + h), radius=radius, fill=fill)


def write_rgb_png(image: Image.Image, path: Path) -> None:
    rgb = image.convert("RGB")
    rgb.save(path, format="PNG")
    data = path.read_bytes()
    if data[:8] != b"\x89PNG\r\n\x1a\n":
        raise SystemExit(f"{path} is not a PNG")
    offset = 8
    while offset + 8 <= len(data):
        length = struct.unpack(">I", data[offset : offset + 4])[0]
        chunk_type = data[offset + 4 : offset + 8]
        chunk = data[offset + 8 : offset + 8 + length]
        if chunk_type == b"IHDR":
            width, height, _bit_depth, color_type = struct.unpack(">IIBB", chunk[:10])
            if (width, height, color_type) != (CANVAS, CANVAS, 2):
                raise SystemExit(f"{path} is {width}x{height} color {color_type}, need {CANVAS} RGB")
            return
        offset += 8 + length + 4
    raise SystemExit(f"{path} has no IHDR")


def main() -> None:
    if len(sys.argv) != 2:
        raise SystemExit(f"usage: {sys.argv[0]} <output.png>")

    image = Image.new("RGB", (CANVAS, CANVAS), FIELD)
    draw = ImageDraw.Draw(image)
    kx, ky, kw, kh = KEYBOARD
    rounded(draw, KEYBOARD, 48, SURFACE)

    row_y = ky + kh - 86
    mid_x = kx + kw / 2
    for row_index, count in enumerate(ROWS):
        row_width = count * KEY_SIZE + (count - 1) * GAP
        x = mid_x - row_width / 2
        if row_index == 2:
            rounded(draw, (x - 72, row_y, 60, KEY_SIZE), 10, SHIFT)
        for _ in range(count):
            rounded(draw, (x, row_y, KEY_SIZE, KEY_SIZE), 10, FIELD)
            x += KEY_SIZE + GAP
        row_y -= KEY_SIZE + 18

    rounded(draw, (mid_x - 160, ky + 36, 320, 44), 10, ACCENT)

    output = Path(sys.argv[1])
    output.parent.mkdir(parents=True, exist_ok=True)
    write_rgb_png(image, output)
    print(f"Wrote {output}")


if __name__ == "__main__":
    main()

File: scripts/generate-icon.swift
#!/usr/bin/env swift
import AppKit

// Colours from https://www.accesstechnology.co.uk
let field = NSColor(srgbRed: 26 / 255, green: 44 / 255, blue: 65 / 255, alpha: 1)
let surface = NSColor(srgbRed: 245 / 255, green: 247 / 255, blue: 250 / 255, alpha: 1)
let shiftColor = NSColor(srgbRed: 36 / 255, green: 118 / 255, blue: 237 / 255, alpha: 1)
let accent = NSColor(srgbRed: 51 / 255, green: 222 / 255, blue: 207 / 255, alpha: 1)

let size = NSSize(width: 1024, height: 1024)
let image = NSImage(size: size, flipped: false) { rect in
    field.setFill()
    rect.fill()

    let keyboardRect = NSRect(x: 168, y: 297, width: 688, height: 430)
    let keyboardPath = NSBezierPath(roundedRect: keyboardRect, xRadius: 48, yRadius: 48)
    surface.setFill()
    keyboardPath.fill()

    let rows = [10, 9, 7]
    let keySize: CGFloat = 48
    let gap: CGFloat = 12
    var y: CGFloat = keyboardRect.maxY - 86
    for (rowIndex, count) in rows.enumerated() {
        let rowWidth = CGFloat(count) * keySize + CGFloat(count - 1) * gap
        var x = keyboardRect.midX - rowWidth / 2
        if rowIndex == 2 {
            let shift = NSRect(x: x - 72, y: y, width: 60, height: keySize)
            shiftColor.setFill()
            NSBezierPath(roundedRect: shift, xRadius: 10, yRadius: 10).fill()
        }
        for _ in 0..<count {
            let keyRect = NSRect(x: x, y: y, width: keySize, height: keySize)
            field.setFill()
            NSBezierPath(roundedRect: keyRect, xRadius: 10, yRadius: 10).fill()
            x += keySize + gap
        }
        y -= keySize + 18
    }

    let space = NSRect(x: keyboardRect.midX - 160, y: keyboardRect.minY + 36, width: 320, height: 44)
    accent.setFill()
    NSBezierPath(roundedRect: space, xRadius: 10, yRadius: 10).fill()

    return true
}

guard let tiff = image.tiffRepresentation,
      let sourced = NSBitmapImageRep(data: tiff) else {
    fputs("Failed to encode PNG\n", stderr)
    exit(1)
}

// App Store Connect rejects 1024 icons that still have an alpha channel,
// even when every pixel is opaque.
guard let flattened = sourced.converting(to: .sRGB, renderingIntent: .default) else {
    fputs("Failed to convert icon to sRGB\n", stderr)
    exit(1)
}
let opaque = NSBitmapImageRep(
    bitmapDataPlanes: nil,
    pixelsWide: flattened.pixelsWide,
    pixelsHigh: flattened.pixelsHigh,
    bitsPerSample: 8,
    samplesPerPixel: 3,
    hasAlpha: false,
    isPlanar: false,
    colorSpaceName: .calibratedRGB,
    bytesPerRow: 0,
    bitsPerPixel: 24
)
guard let opaque else {
    fputs("Failed to allocate opaque icon bitmap\n", stderr)
    exit(1)
}
guard let context = NSGraphicsContext(bitmapImageRep: opaque) else {
    fputs("Failed to draw opaque icon\n", stderr)
    exit(1)
}
NSGraphicsContext.saveGraphicsState()
NSGraphicsContext.current = context
NSImage(cgImage: flattened.cgImage!, size: NSSize(width: flattened.pixelsWide, height: flattened.pixelsHigh))
    .draw(in: NSRect(x: 0, y: 0, width: flattened.pixelsWide, height: flattened.pixelsHigh))
context.flushGraphics()
NSGraphicsContext.restoreGraphicsState()

guard let png = opaque.representation(using: .png, properties: [:]) else {
    fputs("Failed to encode PNG\n", stderr)
    exit(1)
}

let output = URL(fileURLWithPath: CommandLine.arguments[1])
try png.write(to: output)
print("Wrote \(output.path)")
